In [6]:
import numpy as np
import pandas as pd
import tensorflow as tf
from sklearn.preprocessing import LabelEncoder
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

In [7]:
dataset = pd.read_csv('Data/Churn_Modelling.csv')
X = dataset.iloc[:, 3:-1].values
y = dataset.iloc[:, -1].values

# データの前処理

In [ ]:
## カテゴリ変数のエンコーディング
#"Gender"列をエンコーディング
le = LabelEncoder()
X[:, 2] = le.fit_transform(X[:, 2])
#"Geography"列をエンコーディング
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [1])], remainder='passthrough')
X = np.array(ct.fit_transform(X))
# フィーチャースケーリング
sc = StandardScaler()
X = sc.fit_transform(X)
# 訓練データとテストデータに分割
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# ANNの構築

In [10]:
ann = tf.keras.models.Sequential() # Sequentialモデルを作成

In [11]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu')) # 入力層と１つ目の隠れ層を追加

In [12]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu')) # ２つ目の隠れ層を追加

In [13]:
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid')) # 出力層を追加

## 学習方法を設定

In [15]:
ann.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy']) # モデルのコンパイル
# (optimizer='最適化アルゴリズム'、loss='損失関数'、metrics=['評価指標'])

# optimizer: 'adam'は最適化アルゴリズムの一つで、勾配降下法を用いてモデルの重みを更新します。
# loss: 'binary_crossentropy'は2クラス分類のための損失関数で、モデルの予測と実際のラベルとの誤差を計算します。
# metrics: ['accuracy']はモデルの性能を評価するための指標で、正解率を計算します。

# 学習を実行

In [16]:
ann.fit(X_train, y_train, batch_size = 32, epochs = 100)

Epoch 1/100
250/250 [==============================] - 1s 929us/step - loss: 0.7187 - accuracy: 0.5739
Epoch 2/100
250/250 [==============================] - 0s 961us/step - loss: 0.5220 - accuracy: 0.7952
Epoch 3/100
250/250 [==============================] - 0s 929us/step - loss: 0.4473 - accuracy: 0.8067
Epoch 4/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4168 - accuracy: 0.8173
Epoch 5/100
250/250 [==============================] - 0s 1ms/step - loss: 0.3991 - accuracy: 0.8259
Epoch 6/100
250/250 [==============================] - 0s 995us/step - loss: 0.3868 - accuracy: 0.8334
Epoch 7/100
250/250 [==============================] - 0s 908us/step - loss: 0.3774 - accuracy: 0.8394
Epoch 8/100
250/250 [==============================] - 0s 989us/step - loss: 0.3708 - accuracy: 0.8431
Epoch 9/100
250/250 [==============================] - 0s 1ms/step - loss: 0.3659 - accuracy: 0.8445
Epoch 10/100
250/250 [==============================] - 0s 965us/step - loss: 0

# 学習から得た予測、モデルの評価

## シンプルな説明
- **前提**: これは銀行の顧客離れを予測するAIの学習結果
- **目的**: お客さんが銀行を辞めるかどうかを予測する
- **使用データ**: お客さんの年齢、性別、国、残高などの情報
- **結果**: 約87%の確率で正しく予測できるようになった

## わかったこと
- 最初は当て推量レベル（57%）だったが、学習を繰り返すと予測精度が大幅に向上
- 30回ほど学習すると85%まで予測精度が上がり、その後はゆっくり改善
- 2つの隠れ層を持つニューラルネットワークでも高い予測精度が得られた

この人工知能モデルを使えば、銀行は「この人は解約しそうだ」と事前に予測し、対策を打つことができるようになります。